libs/shared/providers/src/wagmi/hooks/useWatchBalances.ts

```diff
@@ -0,0 +1,71 @@
+import { useEffect } from 'react';
+
+import { getTokenId } from '@origin/shared/contracts';
+import { isFulfilled, isNilOrEmpty } from '@origin/shared/utils';
+import { usePrevious } from '@react-hookz/web';
+import { useQuery, useQueryClient } from '@tanstack/react-query';
+import useIdle from 'react-use/lib/useIdle';
+import { mainnet } from 'viem/chains';
+import { useAccount, useBlockNumber, useConfig } from 'wagmi';
+
+import { useWatchBalance } from './useWatchBalance';
+
+import type { Token } from '@origin/shared/contracts';
+import type { HexAddress } from '@origin/shared/utils';
+
+export const useWatchBalances = (args: {
+  tokens: Token[];
+  address?: HexAddress;
+}) => {
+  const config = useConfig();
+  const isIdle = useIdle();
+  const { address } = useAccount();
+  const queryClient = useQueryClient();
+  const addr = args?.address ?? address;
+  const { data: blockNumber } = useBlockNumber({
+    chainId: mainnet.id,
+    watch: true,
+    query: { enabled: !isIdle && !isNilOrEmpty(args.tokens) },
+  });
+  const prev = usePrevious(Number(blockNumber));
+  const res = useQuery({
+    queryKey: [
+      'useWatchBalances',
+      addr,
+      args.tokens?.map((t) => t.symbol),
+      config,
+    ],
+    queryFn: async () => {
+      if (!args.tokens || isNilOrEmpty(args.tokens) || !addr) {
+        return null;
+      }
+
+      let res = {} as Record<string, bigint>;
+
+      const bals = await Promise.allSettled(
+        args.tokens.map((t) =>
+          queryClient.fetchQuery({
```

Comment:
I see what you're trying to do here, keeping a cache of every token balance is a good idea! The best way is to actually mimic the previous wagmi behaviour (can read more about deprecation notice [here](https://wagmi.sh/react/guides/migrate-from-v1-to-v2#deprecated-usebalance-token-parameter)) in a custom useQuery hook, let's call it `useTokenBalance`, wrapping `useReadContract` and `useBalance` (for native/non native distinction) and call this wrapper in our watch hooks with the same mechanism.
This will ensure we keep all balances in the same cache and that they will all be refreshed at the same time!